src/libraries/System.Runtime.Numerics/src/System/Numerics/BigInteger.cs

```diff
@@ -475,9 +475,14 @@ public BigInteger(ReadOnlySpan<byte> value, bool isUnsigned = false, bool isBigE
             AssertValid();
         }
 
-        internal BigInteger(int n, uint[]? rgu)
+        internal BigInteger(int n, uint[]? rgu) : this(n, rgu, false)
         {
-            if ((rgu is not null) && (rgu.Length > MaxLength))
+        }
+
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        private BigInteger(int n, uint[]? rgu, bool bypassBitsLengthCheck)
+        {
+            if (!bypassBitsLengthCheck && (rgu is not null) && (rgu.Length > MaxLength))
```

Comment:
Is this actually meaningful for performance? It feels a bit odd. I imagine you might be asked to revert it

Reply:
The bypassing `.ctor` is (currently) only for `.Abs()` and `op_Neg` which saves a length check for manipulating an existing `BigInteger`, making the resulting JIT code much smaller.

Reply:
This does look overkill. Exceptional path shouldn't matter so much.
Instead, all the callers of the internal constructor should be able to bypass the length check, or do it manually. But this looks overkill too.

Reply:
> This does look overkill. Exceptional path shouldn't matter so much.

Maybe. It's up to your final decisions. Nevertheless, the generated JIT assembly is much smaller, which can be more actively inlined for critical path like `op_Neg`.

<details><summary>code</summary>

```C#
using System;
using System.Diagnostics.CodeAnalysis;
using System.Runtime.CompilerServices;

public readonly struct BigInteger
{
internal static int MaxLength => Array.MaxLength / 32;

    internal readonly int _sign; // Do not rename (binary serialization)
        internal readonly uint[]? _bits; // Do not rename (binary serialization)

         internal BigInteger(int n, uint[]? rgu) : this(n, rgu, false)
        {
        }

        [MethodImpl(MethodImplOptions.AggressiveInlining)]
        private BigInteger(int n, uint[]? rgu, bool bypassBitsLengthCheck)
        {
            if (!bypassBitsLengthCheck && (rgu is not null) && (rgu.Length > MaxLength))
            {
                ThrowOverflowException();
            }

            _sign = n;
            _bits = rgu;
        }

        public static BigInteger OpNeg1(BigInteger value)
        {
            return new BigInteger(-value._sign, value._bits, true);
        }

        public static BigInteger OpNeg2(BigInteger value)
        {
            return new BigInteger(-value._sign, value._bits);
        }

        public static BigInteger Abs1(BigInteger value)
        {
            return new BigInteger(unchecked((int)NumericsHelpers_Abs(value._sign)), value._bits, true);
        }

        public static BigInteger Abs2(BigInteger value)
        {
            return new BigInteger(unchecked((int)NumericsHelpers_Abs(value._sign)), value._bits);
        }

        private static uint NumericsHelpers_Abs(int a)
        {
             unchecked
            {
                uint mask = (uint)(a >> 31);
                return ((uint)a ^ mask) - mask;
            }
        }

        [DoesNotReturn]
        private static void ThrowOverflowException()
        {
            throw new OverflowException();
        }
}
```

</details>

<details><summary>jitted</summary>

```Assembly
BigInteger:OpNeg1(BigInteger):BigInteger (FullOpts):
G_M18931_IG01:  ;; offset=0x0000
G_M18931_IG02:  ;; offset=0x0000
       mov      edx, esi
       neg      edx
       mov      rax, rdi
G_M18931_IG03:  ;; offset=0x0007
       ret      

BigInteger:OpNeg2(BigInteger):BigInteger (FullOpts):
G_M55760_IG01:  ;; offset=0x0000
       push     rbp
       mov      rbp, rsp
G_M55760_IG02:  ;; offset=0x0004
       mov      rax, rdi
       test     rax, rax
       je       SHORT G_M55760_IG04
G_M55760_IG03:  ;; offset=0x000C
       cmp      dword ptr [rax+0x08], 0x3FFFFFE
       jg       SHORT G_M55760_IG06
G_M55760_IG04:  ;; offset=0x0015
       mov      edx, esi
       neg      edx
G_M55760_IG05:  ;; offset=0x0019
       pop      rbp
       ret      
G_M55760_IG06:  ;; offset=0x001B
       call     [BigInteger:ThrowOverflowException()]
       int3     

BigInteger:Abs1(BigInteger):BigInteger (FullOpts):
G_M27312_IG01:  ;; offset=0x0000
G_M27312_IG02:  ;; offset=0x0000
       mov      eax, esi
       sar      eax, 31
       xor      esi, eax
       mov      edx, esi
       sub      edx, eax
       mov      rax, rdi
G_M27312_IG03:  ;; offset=0x000E
       ret      

BigInteger:Abs2(BigInteger):BigInteger (FullOpts):
G_M14547_IG01:  ;; offset=0x0000
       push     rbp
       mov      rbp, rsp
G_M14547_IG02:  ;; offset=0x0004
       mov      eax, esi
       sar      eax, 31
       test     rdi, rdi
       je       SHORT G_M14547_IG04
G_M14547_IG03:  ;; offset=0x000E
       cmp      dword ptr [rdi+0x08], 0x3FFFFFE
       jg       SHORT G_M14547_IG06
G_M14547_IG04:  ;; offset=0x0017
       xor      esi, eax
       mov      edx, esi
       sub      edx, eax
       mov      rax, rdi
G_M14547_IG05:  ;; offset=0x0020
       pop      rbp
       ret      
G_M14547_IG06:  ;; offset=0x0022
       call     [BigInteger:ThrowOverflowException()]
       int3     
```

</details>



> all the callers of the internal constructor should be able to bypass the length check

I would disagree with this, though. During my experience with STJ, parameters for `internal` methods should be checked. It may not be enforced in `BigInteger` but I don't want to make it worse.

Reply:
For internal usage, an assert should be better.

Reply:
In addition, a bool parameter for solely dead code elimination isn't a good practice in this area. Such practice is only used for I/O etc.

If you do want to keep the validation instead of assert in some cases, I'd recommend separating them into two methods.

Reply:
> If you do want to keep the validation instead of assert in some cases, I'd recommend separating them into two methods.

Since they're constructors assigning `readonly` members from identical parameters, they cannot be easily separated. I found the boolean approach most elegant.